handlers/article.go

```diff
@@ -0,0 +1,437 @@
+}
+
+type ArticlesJSON struct {
+	Articles      []Article `json:"articles"`
+	ArticlesCount int       `json:"articlesCount"`
+}
+
+type errorResponse struct {
+	Errors map[string]interface{} `json:"errors"`
+}
+type Author struct {
+	Username  string `json:"username"`
+	Bio       string `json:"bio"`
+	Image     string `json:"image"`
+	Following bool   `json:"following"`
+}
+
+const (
+	contextKeyCurrentUser = contextKey("current_user")
+	contextKeyArticle     = contextKey("article")
+	contextKeyLoggedIn    = contextKey("logged_in")
+)
+
+// ArticlesHandler handle /api/articles
+func (h *Handler) ArticlesHandler(w http.ResponseWriter, r *http.Request) {
+	//h.Logger.Println(r.Method, r.URL.Path)
+
+	// Unprotected routes
+	router := NewRouter(h.Logger)
+	router.AddRoute(
+		`articles\/?$`,
+		"GET", h.getCurrentUser(http.HandlerFunc(h.getArticles)))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"GET", h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.getArticle))))
+
+	// Protected routes
+	router.AddRoute(
+		`articles\/?$`,
+		"POST", h.authorize(h.getCurrentUser(http.HandlerFunc(h.createArticle))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"PUT", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.updateArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"DELETE", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.deleteArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)\/favorite$`,
+		"POST", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.favoriteArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)\/favorite$`,
+		"DELETE", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.unFavoriteArticle)))))
+
+	//router.DebugMode(true)
+
+	router.ServeHTTP(w, r)
+}
+
+func (h *Handler) getCurrentUser(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		var u = &models.User{}
+
+		if claim, _ := h.JWT.CheckRequest(r); claim != nil {
+			u, _ = h.DB.FindUserByUsername(claim.Username)
+		}
+
+		ctx = context.WithValue(ctx, contextKeyCurrentUser, u)
+
+		r = r.WithContext(ctx)
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (h *Handler) extractArticle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		if slug, ok := ctx.Value("slug").(string); ok {
+			a, err := h.DB.GetArticle(slug)
+
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusNotFound)
+				return
+			}
+
+			if a != nil {
+				ctx := r.Context()
+				ctx = context.WithValue(ctx, contextKeyArticle, a)
+				r = r.WithContext(ctx)
+			}
+		}
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (h *Handler) authorize(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if _, err := h.JWT.CheckRequest(r); err != nil {
```

Comment:
Let's get the claims from the `CheckRequest` and add it to the context if its passes